HW9 - 코인헬퍼 Koin 적용

dlwls5201 opened the pull request:
마지막 주차 과제입니다. Repository와 ViewModel을 Koin을 사용해 의존성 주입을 했습니다.
또한 모듈을 사용성에 맞게 분리하여 작성해 보았습니다.

dlwls5201 committed:
HW-9 import koin

dlwls5201 committed:
HW-9 import koin viewmodel

dlwls5201 committed:
HW-9 di 관련 모듈 추가

dlwls5201 committed:
HW-9 applicaion에 startKoin 적용

dlwls5201 committed:
HW-9 MainActivity - DI 적용

dlwls5201 committed:
HW-9 TickerListFragment - DI 적용

preludezdev commented on dlwls5201/MyArchitecture/app/src/main/java/study/architecture/myarchitecture/di/RemoteModule.kt:
```diff
@@ -0,0 +1,28 @@
+            .baseUrl(HOST_URL)
+            .client(get())
+            .addCallAdapterFactory(RxJava2CallAdapterFactory.createAsync())
+            .addConverterFactory(GsonConverterFactory.create())
```
지금은 상관없는데 나중에 여러개 레트로핏 객체 만들 때를 대비해서 안에 인자들도 single 로 만들어도 좋을 것 같아요

dlwls5201 replied:
수정했습니다.

commit : 7d943b7

preludezdev commented on dlwls5201/MyArchitecture/app/src/main/java/study/architecture/myarchitecture/di/RepositoryModule.kt:
```diff
@@ -0,0 +1,14 @@
+val repositoryModule = module {
+
+    single<UpbitRepository> {
+        UpbitRepositoryImpl(get(), get(), NetworkUtil.isOnline(androidApplication()))
```
위에 코드랑은 상관없는데 혹시 NetworkUtil 클래스 안에 있는
`val networkInfo: NetworkInfo? = connMgr.activeNetworkInfo`  부분에서
null 로 받게 되는 경우가 있나요?

dlwls5201 replied:
자바 코드를 복붙해서 null 체크 코드가 되었네요. null 여부를 제거해도 경고메시지가 없는걸로 봐서는 null 로 받는 경우는 없는것 같습니다 :)

commit : 4506805

preludezdev reviewed (commented):
역시 피드백 드릴게 별로 없네요.
뷰모델에 인자 넣는거나 필터링기능 추가 등등 진성님 코드 보고 많이 배웠습니다 : )

P.S. 이전에 올린 PR은 close 하셔도 될거같아요~

dlwls5201 committed:
HW-9 Network 모듈 수정

dlwls5201 committed:
HW-9 Room Warning 제거

dlwls5201 committed:
HW-9 NetworkUtil null 체크 없앰

dlwls5201 committed:
HW-9 BaseFragment lifecycleOwner 변경

preludezdev commented:
Approve 는 멘토 리뷰 마치면 하도록 하겠습니다!

BeokBeok commented on dlwls5201/MyArchitecture/app/build.gradle:
```diff
@@ -42,6 +42,10 @@ dependencies {
 
+    //koin
+    implementation "org.koin:koin-android:$koinVersion"
+    implementation "org.koin:koin-android-viewmodel:$koinVersion"
```
androidx 를 쓰신다면 androidx 용 koin 을 쓰시는 것이 좋을 것 같아요.
```suggestion
    implementation "org.koin:koin-androidx-viewmodel:$koinVersion"
```

dlwls5201 replied:
수정 및 적용 했습니다.

commit : 920db97 , 30119ea

companion object 제거는 LocalDatabase만 적용했고 나머지 객체는 사용하지 않기 때문에 코드에서 제거했습니다.

commit : 8f6d030

zozero94 replied:
질문이 있습니다.
- companion object를 사용하여 객체를 만드는 부분(SIngleTon)
을 제거하라 하신 이유가 koin에서 내부적으로 싱글톤으로 생성해주는데
한번 더 하는 불필요한 작업이기때문에 그런가요?

- 그렇다면 만약 나중에 협업을 할 때 누군가가 그 클래스를 koin을 사용하지 않고
직접 만들어 사용하게 된다면 문제가 발생하지 않나요??

BeokBeok replied:
> 질문이 있습니다.
> 
> * companion object를 사용하여 객체를 만드는 부분(SIngleTon)
>   을 제거하라 하신 이유가 koin에서 내부적으로 싱글톤으로 생성해주는데
>   한번 더 하는 불필요한 작업이기때문에 그런가요?

A : 본 프로젝트에 작성된 companion object 는 
외부에서 객체를 생성하기 위해서 만든 것으로 알고 있습니다.
하지만, koin 라이브러리를 통하여 사용하게 된다면, 
의존성 주입을 koin 이 해주기 때문에
companion object 에 정의된 함수를 통하여 객체를 생성하는 것은 의미가 없다고 봅니다.

> * 그렇다면 만약 나중에 협업을 할 때 누군가가 그 클래스를 koin을 사용하지 않고
>   직접 만들어 사용하게 된다면 문제가 발생하지 않나요??

A : 말씀하신 부분에 대해서는 개발자끼리 협의해서 
koin 을 적용한 프로젝트에서는 companion object 를 사용하여 객체를 생성하는 코드를 작성하지 않도록 해야 할 것으로 판단됩니다.
하지만, koin 을 사용하기 때문에 companion object 를 사용하면 안된다 라는 의미는 아닙니다.

BeokBeok reviewed (changes requested):
안녕하세요. 스터디 2기 강현석이라고 합니다.
부족하지만 잘 부탁 드립니다. 😅 

---

koin 을 사용하여 의존성 주입을 하는 경우에는
기존에 companion object 를 사용하여 객체를 만드는 부분은 제거하시는 것이 좋습니다.

- LocalDataBase 의 companion object
- RemoteClient object
- Injection object
- MainViewModelFactory
- TickerListViewModelFactory

preludezdev commented:
멘토라벨이 없어진김에... 
mainAdapter에서 fragments 리스트를 WeakReference 로 만드셨던데 그냥 만들게 되면 메모리 leak 이 발생하는 구간이 있나요? 혹시 있다면 어느 부분인지 설명해줄 수 있으신가요?

preludezdev commented on dlwls5201/MyArchitecture/app/src/main/java/study/architecture/myarchitecture/ui/main/MainActivity.kt:
```diff
@@ -2,22 +2,15 @@ package study.architecture.myarchitecture.ui.main
+    private val mainViewModel: MainViewModel by viewModel()
 
     private val mainAdapter: MainAdapter by lazy {
         MainAdapter(supportFragmentManager)
```
- mainAdapter에서 fragments 리스트를 WeakReference 로 만드셨던데 그냥 만들게 되면 메모리 leak 이 발생하는 구간이 있나요? 혹시 있다면 어느 부분인지 설명해줄 수 있으신가요? 
- WeakReference 가 어떻게 쓰이는지 찾아보니까 GC 에 collect 되는걸 강화(?) 시키는 느낌이던데 만약에 해당 fragment 가 GC 에 collect 되서 null 이 되면 나중에 다시 fragment 가져오는 부분에서 대처해줘야 하지 않나요?

dlwls5201 replied:
Fragment 사용은 어렵네요. 저도 이런 저런 테스트를 해봐야 될 것 같습니다. 이 부분은 https://code-examples.net/ko/q/6a721b 글을 참고해서 만들었습니다.

프래그먼트 조각을 재사용하기 위해 위와 같이 적용했으면 Activity 디자인에 따라 프래그먼트가 존재할 수 도 있고 없을 수도 있어 WeakReference 사용을 권하는 글이었습니다. 실재로 프래그먼트 조각들의 참조가 어떻게 동작되는지는 이런 저런 시도를 해봐야 알 수 있을 것 같네요:)

- 두번 째 질문에서 fragmentPagetAdapter가 getItem에서 프래그먼트를 생성하고 instantiateItem에서 참조된 프래그먼트를 불러옵니다. 프래그먼트가 GC에 의해 null이되면 자동으로 getItem이 호출되어 프래그먼트가 생성되는 구조일것 같은데...확신이 없네요 ㅎㅎ

preludezdev reviewed (commented):
궁금증이 남아 리뷰남깁니다!!

ps. 실수로 Close Pull request 눌렀어요ㅠㅠ

dlwls5201 committed:
HW-9 koin-androidx-viewmodel 수정 적용

dlwls5201 committed:
HW-9 import androidx.viewmodel

dlwls5201 committed:
HW-9
 - DI를 적용해 LocalDataBase 생성
 - 불필요한 파일 4개 제거
 (RemoteClient, Injection. MainViewModelFactory, TickerListViewModelFactory)

BeokBeok reviewed (approved):
LGTM:ok_hand:

sjjeong reviewed (approved):
LGTM ✊ ✌️ 👋

sjjeong commented:
지금까지 고생하셨습니다~
후기 남겨주세요!

dlwls5201 commented:
우와~! 감사합니다~! 후기는 차차 작성하겠습니다 :)
